feat(docs-infra): add useful links if landed on 404 page

ajitsinghkaler opened the pull request:
Added additional links useful links which can help user find the things he is looking for on the 404 page
Fixes #31532

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying or link to a relevant issue. -->
On no link found no additional shown which can be helpful

Issue Number: 31532


## What is the new behavior?
Added additional links which can be helpful to users to find what they are looking for

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gkalpak reviewed (changes requested):
Thx for taking this on, @ajitsinghkaler! Per the discussion in #31532, the generic links would appear if there can be no relevant search results. (I would be fine putting them below the search result too.)

So, they need to be somehow incorparated into the `aio-file-not-found-search` component.

ajitsinghkaler commented:
@gkalpak I've made the changes please have a look

ajitsinghkaler commented:
@gkalpak I've made the changes please have a look I've incorporated it in the aio-file-not-found-search component.

ajitsinghkaler commented:
@gkalpak can you please explain why e2e tests are failing

ajitsinghkaler commented:
@gkalpak can you please look into this too

mary-poppins commented:
You can preview 99c2faf at https://pr34978-99c2faf.ngbuilds.io/.
You can preview 19aee04 at https://pr34978-19aee04.ngbuilds.io/.
You can preview e4c6619 at https://pr34978-e4c6619.ngbuilds.io/.

gkalpak reviewed (changes requested):
Regarding why the e2e tests are failing:

The `SearchResultsComponent` will initially have an empty result set (since it hasn't received the search results yet) and will therefore show the [#notFound template][1]. At that point, however, the `notFoundMessage` property is [set to "Searching..."][2].

Then, once the results are received, the `notFoundMessage` is [changed to "No results found"][3] and either the [#searchResults][4] or the [#notFound][1] templates are shown (the latter now showing "No results found" instead of "Searching...").

So, while we pretend there are two states (no results or results), there are actually three: searching, no results, results
But we are piggybacking on the "no results" state to handle the "searching" case too.

##
Now, why this causes the e2e to fail?

In the e2e tests, we [send a search query][5] and run assertions against the [search results][6]. The way we retrieve the search results is by [waiting][7] for [elements that match `.search-results li`][8].

Previously, such elements would only be present when there were some actual search results. With your changes, these elements are also present whenever the [#notFound][1] template is shown, which (as explained above) is also showing while the search is in progress.

As a result, the e2e tests will look at the generic links and think they are the search results returned by the search query and thus try to run assertions on them. These assertions will fail, of course, because these "results" do not contain the expected terms.

---

**How to fix this?**

I think the proper fix is to have three templates:
- One for when there are results (i.e. the `#searchResults` template we already have).
- One for when there are no results showing the generic links (i.e. the `#notFound` template, except with the `notFoundMessage` hard-coded to "No results found.").
- One for when the search is in progress without the generic links (and with a hard-coded "Searching..." message).

Then we would also need a property on the component to track whether a search is in progress or not.

LMK if that makes sense and whether you need any help with the implementation.

[1]: https://github.com/angular/angular/blob/c35671c0a/aio/src/app/shared/search-results/search-results.component.html#L28-L30
[2]: https://github.com/angular/angular/blob/c35671c0a/aio/src/app/shared/search-results/search-results.component.ts#L26
[3]: https://github.com/angular/angular/blob/c35671c0a/aio/src/app/shared/search-results/search-results.component.ts#L46
[4]: https://github.com/angular/angular/blob/c35671c0a/aio/src/app/shared/search-results/search-results.component.html#L5-L26

[5]: https://github.com/angular/angular/blob/c35671c0a/aio/tests/e2e/src/search.e2e-spec.ts#L12
[6]: https://github.com/angular/angular/blob/c35671c0a/aio/tests/e2e/src/search.e2e-spec.ts#L13
[7]: https://github.com/angular/angular/blob/c35671c0a/aio/tests/e2e/src/app.po.ts#L83
[8]: https://github.com/angular/angular/blob/c35671c0a/aio/tests/e2e/src/app.po.ts#L82

ajitsinghkaler commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -26,5 +26,25 @@ <h3 class="search-section-header">{{area.name}} ({{area.pages.length + area.prio
+        </li>
+        <li  class="search-page">
+            <a  class="search-result-item" href="https://blog.angular.io/">Angular blog</a>
+        </li>
```
YS WILL DO

ajitsinghkaler commented on aio/src/styles/2-modules/_search-results.scss:
```diff
@@ -103,6 +111,14 @@ aio-search-results {
+        flex-direction: column;
+        a {
+          color: $blue;
+          font-weight: 500;
```
there we applied color white which is not visible when you enter a wrong URL on the aio site so that is why I applied color blue as for font-weight I'll remove it.

gkalpak replied:
Yes, I was referring to the `font-weigth` line specifically. Thx!

mary-poppins commented:
You can preview 4da55d9 at https://pr34978-4da55d9.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak I've made the changes pleas have a look

mary-poppins commented:
You can preview 6076979 at https://pr34978-6076979.ngbuilds.io/.

mary-poppins commented:
You can preview fed6345 at https://pr34978-fed6345.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak  why these hooks are failing can you please explain this.

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
 <div class="search-results">
-  <div *ngIf="searchAreas.length; then searchResults; else notFound"></div>
-</div>
+  <div>
```
This extra `div` seems unnecessary. Can we get rid of it (or does it serve a purpose I missed)?

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
-  <div *ngIf="searchAreas.length; then searchResults; else notFound"></div>
-</div>
+  <div>
+    <div [ngSwitch]="searchState">
```
Actually, we shouldn't need this `div` either. Can you please move the `ngSwitch` to the top-level `div` instead?

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
+  <div>
+    <div [ngSwitch]="searchState">
+
+      <div *ngSwitchCase="1">
```
Let's avoid unnecessary DOM elements (here and below):

```suggestion
      <ng-container *ngSwitchCase="1">
```

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
+    <div [ngSwitch]="searchState">
+
+      <div *ngSwitchCase="1">
+        <p class="not-found">Searching ...</p>
```
Let's change the class name to something more generic (e.g. `no-results`). `not-found` implies that we searched for andf couldn't find them, which is not the case here.

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
+      </div>
+      
+      <div *ngSwitchCase="3">
+        <div>
```
Again, let's avoid this unnecessary `div` and convert the parent `div` to `ng-container`.

gkalpak commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,59 @@
+      <div *ngSwitchCase="3">
+        <div>
+          <p class="not-found">No results found.</p>
+          <p class="not-found">Here are a few links that might be helpful in finding what you are looking for:</p>
```
Wrap these two elements in a parent element with `class="..."` (instead of having the class on each element (just to set their color)).

gkalpak commented on aio/src/app/shared/search-results/search-results.component.ts:
```diff
@@ -23,11 +23,24 @@ export class SearchResultsComponent implements OnChanges {
   ngOnChanges() {
+    if (this.searchResults == null) {
+      this.searchState = '1';
+    } else if (this.searchResults['results'].length) {
```
```suggestion
    } else if (this.searchResults.results.length) {
```

gkalpak commented on aio/src/app/shared/search-results/search-results.component.ts:
```diff
@@ -23,11 +23,24 @@ export class SearchResultsComponent implements OnChanges {
+   * 2 is results found state
+   * 3 or greater is not found state
+   */
+  searchState = '1';
```
Let's use a string enum instead, so that the values are more descriptive (and we get better type-checking):

```ts
enum SearchState {
  InProgress = 'in-progress',
  ResultsFound = 'results-found',
  NoResultsFound = 'no-results-found'
}
```

gkalpak reviewed (changes requested):
Thx for implementing this, @ajitsinghkaler! It is much cleaner now.
Can you also update the unit tests to at least verify when the extra links are shown?

mary-poppins commented:
You can preview a750bd8 at https://pr34978-a750bd8.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak can you please review this I've made the changes

mary-poppins commented:
You can preview 9ebc0e9 at https://pr34978-9ebc0e9.ngbuilds.io/.

gkalpak commented on aio/src/app/shared/search-results/search-results.component.spec.ts:
```diff
@@ -214,9 +216,39 @@ describe('SearchResultsComponent', () => {
+
+    it('should contain blog reference link', () => {
+      expect(getHrefs(anchors)[4]).toContain('https://blog.angular.io/');
+    })
```
Let's combine all these tests into one:

```ts
const resultLinks = fixture.debugElement.queryAll(By.css('.search-page a'));
const resultHrefs = resultLinks.map(a => a.nativeNode.href);
expect(resultHrefs).toEqual([...]);
```

gkalpak commented on aio/src/app/shared/search-results/search-results.component.spec.ts:
```diff
@@ -214,9 +216,39 @@ describe('SearchResultsComponent', () => {
 
   describe('when no query results', () => {
-    it('should display "not found" message', () => {
+    let anchors: DebugElement[]; 
```
Since this will only be used on one test (see my other comment), we don't need to define it here.
Same for the `getHrefs()` helper.

gkalpak reviewed (changes requested):
Thx for the tests. Can you also add some tests to verify that the generic links are not shown in the other states (i.e. while searching is in progress and when there are results found)?

Also, please, make sure each commit message follows the [commit message guidelines][1] and represents a self-contained unit of work (e.g. tests should be on the same commit as the code changes that are tested).

[1]: https://github.com/angular/angular/blob/939e2f76f/CONTRIBUTING.md#commit

ajitsinghkaler commented:
@gkalpak I 

> Thx for the tests. Can you also add some tests to verify that the generic links are not shown in the other states (i.e. while searching is in progress and when there are results found)?
> 
> Also, please, make sure each commit message follows the [commit message guidelines](https://github.com/angular/angular/blob/939e2f76f/CONTRIBUTING.md#commit) and represents a self-contained unit of work (e.g. tests should be on the same commit as the code changes that are tested).

I'm just pushing changes right now when we get this finalized I'll squash all the commits into one until them I'm pushing new changes on this branch if that is not much of a problem. Otherwise, I can squash them right now.

gkalpak commented:
That's OK :+1:

ajitsinghkaler commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,57 @@
+<div class="search-results" [ngSwitch]="searchState">
+
+  <ng-container *ngSwitchCase="'in-progress'">
+    <p class="not-found">Searching ...</p>
```
Done

ajitsinghkaler commented on aio/src/app/shared/search-results/search-results.component.html:
```diff
@@ -1,30 +1,57 @@
+  </ng-container>
+  
+  <ng-container *ngSwitchCase="'no-results-found'">
+    <div>
```
Yes this is needed if we remove this div the p tag and the ul tag comes side by rather than one below anther because of search results flex

gkalpak replied:
I see. Then, how about adding the `search-area` class to it?

ajitsinghkaler commented:
@gkalpak made the changes please have a look

gkalpak commented on aio/src/app/shared/search-results/search-results.component.spec.ts:
```diff
@@ -216,39 +220,19 @@ describe('SearchResultsComponent', () => {
+      const resultLinks = fixture.debugElement.queryAll(By.css('.search-page a'));
+      const resultHrefs = resultLinks.map(a => a.nativeNode.href);
+      expect(resultHrefs.length).toEqual(5);
+      expect(resultHrefs[4]).toEqual("https://blog.angular.io/");
```
Why only test one entry? Let's test all:

```ts
expect(resultHrefs).toEqual([
  ...entry 1...
  ...entry 1...
  ...
]);
```

ajitsinghkaler replied:
How do I get base href of the app

gkalpak replied:
Hm...you can just use `a.nativeNode.getAttribute('href')` and check the `href` attributes (which is what we set in the HTML, not the full resolved URL.

ajitsinghkaler commented:
@gkalpak can you please look at it I made the changes

ajitsinghkaler commented:
@gkalpak can you please look at it I made the changes

gkalpak commented on aio/src/app/shared/search-results/search-results.component.spec.ts:
```diff
@@ -230,9 +230,15 @@ describe('SearchResultsComponent', () => {
+        'resources', 
+        'guide/glossary', 
+        'guide/cheatsheet', 
+        'https://blog.angular.io/'
```
Nit:
```suggestion
        'https://blog.angular.io/',
```

gkalpak reviewed (approved):
Thx, @ajitsinghkaler! LGTM - don't forget to squash commits.

ajitsinghkaler committed:
feat(docs-infra): add useful links if landed on 404 page and no search results found

Added additional links which can help user find the things they are
looking for when there are no search results (when searching or on a 404
page).

Note:
This commit increases the main bundle's payload size due to the extra
content of the `aio-search-results` component.

Fixes #31532

ajitsinghkaler commented:
@gkalpak made all the changes and did the cleanup please have a look

mary-poppins commented:
You can preview 020cd88 at https://pr34978-020cd88.ngbuilds.io/.

gkalpak commented:
Hm...this PR seems to increase the main bundle by 1KB. Investigating whether that's expected...

gkalpak committed:
ci(docs-infra): increase AIO ViewEngine payload size limit

In #35702, the payload size limit for Ivy builds was bumped to account
for small incremental increases in recent PRs. The ViewEngine size has
also increased similarly (~500B), but it was not updated in #35702,
because its total increase was just below the 500B error threshold (by
6B).

This commit bumps the ViewEngine size limit too.

Note: Any investigation for the Ivy size increase (as a follow-up
to #35702) will most likely also apply to ViewEngine, since the size was
increased by the same amount.

mary-poppins commented:
You can preview 57563b0 at https://pr34978-57563b0.ngbuilds.io/.

gkalpak committed:
fixup! feat(docs-infra): add useful links if landed on 404 page and no search results found

gkalpak commented:
@googlebot I consent.

gkalpak commented:
OK, it seems that the increase is expected (and is caused by the changes and extra content in the `aio-search-results` component). For Ivy, it is ~1KB (uncompressed).

I rebased on master, updated payload size limits for current master (to make it easy to see the increase by this PR) and bumped the payload size limits due to this PR's changes.

mary-poppins commented:
You can preview 383a277 at https://pr34978-383a277.ngbuilds.io/.
